P4RT-5.3: Traceroute: PacketIn With VRF Selection changes incorporated

ampattan opened the pull request:
- Updated `metadata.textproto` to include vendor check for NOKIA and added deviation.
- Fixed issues in `traceroute_packetin_with_vrf_selection_test.go`.
- Implemented modifications in the `vrfpolicy.go` file.

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

ampattan committed:
P4RT-5.3: Traceroute: PacketIn With VRF Selection changes incorporated

- Updated metadata.textproto to include vendor check for NOKIA and added deviation.
- Fixed issues in traceroute_packetin_with_vrf_selection_test.go.
- Implemented modifications in the vrfpolicy.go file.

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

nsadhasivam commented:
/fptest virtual

nsadhasivam commented:
/fptest physical

ram-mac commented:
@ampattan - I have validated the test and the test is failing. Test was run on 7250 IXR-18e with image 0.0.1-75140
Can you please take a look into it. I will share the logs

ampattan commented:
> @ampattan - I have validated the test and the test is failing. Test was run on 7250 IXR-18e with image 0.0.1-75140 Can you please take a look into it. I will share the logs

@ram-mac I have validated the test on the specified build and platform and did not encounter the issue. Based on the logs you shared, it appears to be a setup-related problem and not related to the test case itself. Please review the setup configuration. I will also share the logs for reference.

If you encountered any sub-testcase failures, yes — a couple of sub-testcases from P4rt-5.3 (specifically TC-6 and TC-9) are expected to fail on the given build for the J3(IXR-18e/10e) platform due to a known SRL issue. This issue has already been identified and fixed, and the fixes will be available in the next build drop.

ram-mac commented:
> > @ampattan - I have validated the test and the test is failing. Test was run on 7250 IXR-18e with image 0.0.1-75140 Can you please take a look into it. I will share the logs
> 
> @ram-mac I have validated the test on the specified build and platform and did not encounter the issue. Based on the logs you shared, it appears to be a setup-related problem and not related to the test case itself. Please review the setup configuration. I will also share the logs for reference.
> 
> If you encountered any sub-testcase failures, yes — a couple of sub-testcases from P4rt-5.3 (specifically TC-6 and TC-9) are expected to fail on the given build for the J3(IXR-18e/10e) platform due to a known SRL issue. This issue has already been identified and fixed, and the fixes will be available in the next build drop.

@ampattan - There is no issue with the setup. The test is failing due to unsupported leaf. See below
        }: Schema '/interfaces/interface/ethernet/config' has no local leaf with the name 'auto-negotiate'. Options are [mac-address, port-speed, aggregate-id]: Parse error on line 4:       "auto-negotiate": false,
        Input:
        '{
          "openconfig-if-ethernet:ethernet": {
            "config": {
              "auto-negotiate": false,
              "duplex-mode": "FULL",
              "port-speed": "openconfig-if-ethernet:SPEED_100GB"
            }
          },

Please fix it and let me know.

ampattan committed:
Merge remote-tracking branch 'upstream/main' into AP-p4rt5dot3

ampattan committed:
Comment addressed reverted the isis changes.

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

ampattan commented on feature/p4rt/otg_tests/traceroute_packetin_with_vrf_selection_test/traceroute_packetin_with_vrf_selection_test.go:
```diff
@@ -742,10 +746,10 @@ func configureISIS(t *testing.T, dut *ondatra.DUTDevice, intfName, dutAreaAddres
+		intfName = intfName + ".0"
+	}
 	isisIntf := isis.GetOrCreateInterface(intfName)
-	isisIntf.GetOrCreateInterfaceRef().Interface = ygot.String(intfName)
```
@ram-mac comment addressed, restored the changes.

bkreddy143 commented:
@ram-mac @ampattan  ,  'PMD_100GBASE_FR'  suppose to used fo 100G port with FR optics installed. As per the logs DUT has ports are in nx100G mode and it is not actually 100G optics so we shouldnt set pmd for such ports.

ampattan committed:
Merge branch 'main' into AP-p4rt5dot3

ram-mac commented on internal/vrfpolicy/vrfpolicy.go:
```diff
@@ -362,7 +362,7 @@ func buildVRFSelectionPolicy(niName string, policyName string, pfRules []*policy
 			}
 		} else {
-			pfRProtoIP := pfR.GetOrCreateIpv4()
+			pfRProtoIP := pfR.GetOrCreateIpv6()
```
@ampattan - This change is not required, the latest code already has the corrected one

ram-mac commented on feature/p4rt/otg_tests/traceroute_packetin_with_vrf_selection_test/traceroute_packetin_with_vrf_selection_test.go:
```diff
@@ -867,7 +868,7 @@ func configureOTG(t testing.TB, otg *otg.OTG, atePorts []*ondatra.Port) gosnappi
 		port := config.Ports().Add().SetName(ap.ID())
 		atePortNamelist = append(atePortNamelist, port.Name())
-		portName := fmt.Sprintf("atePort%s", strconv.Itoa(i+1))
+		portName := fmt.Sprintf("ate%s", strings.ToUpper(string(ap.ID()[0]))+string(ap.ID()[1:]))
```
@ampattan  - We are moving away from the strconv to strings here; What is the issue with strconv. And i think there is below change in the condition removing if i == 7 and moving to if ap.ID() == "port8"

ampattan replied:
Comment addressed.

ram-mac commented on feature/p4rt/otg_tests/traceroute_packetin_with_vrf_selection_test/traceroute_packetin_with_vrf_selection_test.go:
```diff
@@ -1272,25 +1271,26 @@ func TestTraceRoute(t *testing.T) {
 		testTunnelTrafficDecapEncap(ctx, t, dut, args)
 	})
+
+	defer func() {
```
@ampattan - Why are we adding this defer and clean up function? In general there is a cleanup process after the test is run.

ampattan replied:
Removed defer function.

ram-mac commented:
@ampattan - Also please fix the static analysis errors in this PR

ampattan committed:
Merge branch 'main' into AP-p4rt5dot3

ampattan committed:
Merge branch 'main' into AP-p4rt5dot3

ampattan commented:
> @ampattan - Also please fix the static analysis errors in this PR

Static analysis errors are no longer being displayed.

ampattan commented:
> @ram-mac @ampattan , 'PMD_100GBASE_FR' suppose to used fo 100G port with FR optics installed. As per the logs DUT has ports are in nx100G mode and it is not actually 100G optics so we shouldnt set pmd for such ports.

@ram-mac Wrt https://issuetracker.google.com/424761469. it has been addressed here.

ampattan committed:
Merge branch 'main' into AP-p4rt5dot3

ampattan committed:
PR comments addressed

AmrNJ reviewed (approved):
There needs to be a change in packetin_test.go to avoid flaky failures. I will fix it in a different PR. LGTM.
